src/Runtime/OMTensorList.inc

```diff
@@ -51,60 +63,91 @@ struct OMTensorList {
    *
    * Destroy the OMTensorList struct.
    */
-  ~OMTensorList() {
-    /* Destroy all the OMTensors */
-    for (int i = 0; i < _size; i++)
-      if (_omts[i])
-        omTensorDestroy(_omts[i]);
-  };
+  ~OMTensorList() { cleanup(this); };
 #endif
 
   /* To facilitate user facing API getOmts, OMTensors are kept in a vector
    * that can be quickly returned as an array. A name to index map is used
    * to address ReMemRefs by name.
    */
   OMTensor **_omts; // OMTensor array
-
-  size_t _size; // Number of elements in _omts.
-
-
-  int _owning; // indicates whether the OMTensorList owns the pointer to
+  size_t _size;     // Number of elements in _omts.
+  int _owning;      // indicates whether the OMTensorList owns the pointer to
                // OMTensor array or not. OMTensorList struct will release the
                // memory space referred to by '_omts' upon destruction if and
                // only if it owns it.
-
+  OMTensor *_internalOmts[OMTENSOR_SUPPORTED_INTERNAL_SIZE];
 };
 
+// Cleanup the list, which include removing all tensors present in the list.
+// With reset on, the fields will be reset (safer if used after cleanup).
+static void cleanup(OMTensorList *list) {
+  for (int i = 0; i < list->_size; i++)
+    if (list->_omts[i])
+      omTensorDestroy(list->_omts[i]);
+  if (list->_owning || list->_size > OMTENSOR_SUPPORTED_INTERNAL_SIZE)
+    free(list->_omts);
+}
+
+// Initialize an OMTensorList with mandatory size and list of tensors. Return 0
+// on failure, and cleanup the data structure.
+static int init(OMTensorList *list, size_t size, int owning, OMTensor **omts) {
+  assert(list);
+  if (size > 0)
+    assert(omts);
+  // Initialize fields.
+  list->_size = size;
+  list->_owning = owning;
+  if (owning) {
+    // no copying, just reuse the list
+    list->_omts = omts;
```

Comment:
I think you have it backwards. For both `OMTensor`/`OMTensorList` `_owning=true` means that we own the memory of the tensor data/list so cleanup should free the memory. Otherwise, cleanup does not free the memory. So `_owning=true` is for the case where you copied (and own) the list and need to free it.

Reply:
I want to change the behavior. I think it is bad form to force the user to keep the list of ptr "alive" for the duration of the live of the list of tensors.

We don't do it for shape and strides... there we just copy the lists. We should do the same for the list of tensors. I would like to remove the "owning" entirely of the interface, but it is only used by the compiler (function is not in the runtime public header file), so I am willing to ignore it.

Reply:
The meaning of `_owning` for OMTensor and OMTensorList should be consistent. Otherwise it's going to be confusing since the devs will see both public and private APIs.